e2e/inspect/inspect.go

```diff
@@ -0,0 +1,182 @@
+			"org.label-schema.usage.singularity.version": "3.2.1-660.g4c8a84050"
+		}
+	},
+	"type": "container"
+}`
+
+const expectedRunscriptJson = `
+{
+	"attributes": {
+		"apps": "",
+		"runscript": "#!/bin/sh\n\necho \"E2E Test Container\"\n\n\n"
+	},
+	"type": "container"
+}`
+
+const expectedListAppsJson = `
+{
+	"attributes": {
+		"apps": "hello\nworld\n"
+	},
+	"type": "container"
+}`
+
+const expectedTestJson = `
+{
+	"attributes": {
+		"apps": "",
+		"test": "#!/bin/sh\n\nls /\ntest -d /\ntest -d /etc\n\n\n"
+	},
+	"type": "container"
+}`
+
+const expectedEnvironmentJson = `
+{
+	"attributes": {
+		"apps": "",
+		"environment": {
+			"90-environment.sh": "#!/bin/sh\n#Custom environment shell code should follow\n\n\nexport test=\"testing\"\nexport e2e=\"e2e testing\"\n\n\n"
+		}
+	},
+	"type": "container"
+}`
+
+func runInspectCommand(inspectType string) ([]byte, error) {
+	argv := []string{"inspect", "--json", inspectType, "testdata/test.sif"}
+	cmd := exec.Command("singularity", argv...)
+
+	return cmd.CombinedOutput()
+}
+
+func singularityInspect(t *testing.T) {
+	tests := []struct {
+		name      string
+		insType   string
+		json      []string
+		expectOut string
+	}{
+		{
+			name:      "label E2E",
+			insType:   "--labels",
+			json:      []string{"attributes", "labels", "E2E"},
+			expectOut: expectedLabelsJson,
+		},
+		{
+			name:      "label",
+			insType:   "--labels",
+			json:      []string{"attributes", "labels", "HI"},
+			expectOut: expectedLabelsJson,
+		},
+		{
+			name:      "label",
+			insType:   "--labels",
+			json:      []string{"attributes", "labels", "e2e"},
+			expectOut: expectedLabelsJson,
+		},
+		{
+			name:      "label",
+			insType:   "--labels",
+			json:      []string{"attributes", "labels", "hi"},
+			expectOut: expectedLabelsJson,
+		},
+		{
+			name:      "runscript",
+			insType:   "--runscript",
+			json:      []string{"attributes", "runscript"},
+			expectOut: expectedRunscriptJson,
+		},
+		{
+			name:      "list apps",
+			insType:   "--list-apps",
+			json:      []string{"attributes", "apps"},
+			expectOut: expectedListAppsJson,
+		},
+		{
+			name:      "test",
+			insType:   "--test",
+			json:      []string{"attributes", "test"},
+			expectOut: expectedTestJson,
+		},
+		{
+			name:      "environment",
+			insType:   "--environment",
+			json:      []string{"attributes", "environment", "90-environment.sh"},
+			expectOut: expectedEnvironmentJson,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, test.WithPrivilege(func(t *testing.T) {
+			out, err := runInspectCommand(tt.insType)
+			if err != nil {
+				t.Fatalf("Unexpected failure: %s: %s", string(out), err)
+			}
+
+			// Check the E2E label in test.sif, does it match our expected output
+			v, err := jsonparser.GetString(out, tt.json...)
+			if err != nil {
+				fmt.Println("ERROR: ", err)
+			}
+			// Get the expected output, and compair them
+			e, err := jsonparser.GetString([]byte(tt.expectOut), tt.json...)
```

Comment:
but this bit here is extracting the same object from the stored dump. This seems strange. This is where I would have expected to compare v above with a specific value.

Reply:
This is just getting the expected out (`tt.expectOut`). Then is comparing it on line ~177

Reply:
I guess we _could_ have a parameter in the table tests setting the expected output.

But, why make it simple, when you can make more complex and hard to read? Lol

Reply:
the bit that is strange about this is that you have a whole JSON object stored in `expectOut`, but you are looking at a specific value within that object for the test.

I think I know what's bugging me here.

The code is checking that each one of the labels is correct, and it has one test case per label. At a higher level, what you want to do is check that all the labels are present, and that their values are correct. That's what you have in `expectedLabelsJSON`, the whole set of expected labels.

The problem with comparing JSON object as strings is that nothing says their order should be the same. For example `{"foo": 1, "bar": 2}` is the same object as `{"bar": 2, "foo": 1}`, even if those are different strings.

If you want two compare two JSON objects represented as strings, you can do [this](https://play.golang.org/p/zD81n-nX7uI). In this way you can have a single test case for each `--labels`, `--runscript`, etc.

Reply:
I'll go back to your first suggestion, just adding the expected output in the table.